O objetivo desse Jupyter é servir como um guia para mais pessoas que querem aprender como modelar os dados e analisar seguindo o Python, passarmos de análises superficiais para modelos preditivos e Machine Learning e a partir disso tirarmos alguns insights interessantes.

Vamos iniciar importando a biblioteca Pandas para que possamos tratar os dados.

In [1]:
import pandas as pd

Agora podemos visualizar o dataset utilizando o Pandas.
Esse dataset foi disponibilizado no repositório do github.

In [2]:
df = pd.read_excel('default_of_credit_card_clients__courseware_version_1_21_19.xlsx')

Visualizando as 5 primeiras linhas do dataset:

In [3]:
df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0  798fc410-45c1      20000    2          2         1   24     2      2   
1  8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2  85698822-43f5      90000    2          2         2   34     0      0   
3  0737c11b-be42      50000    2          2         1   37     0      0   
4  3b7f77cc-dbc0      50000    1          2         1   57    -1      0   

   PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0     -1     -1  ...          0          0          0         0       689   
1      0      0  ...       3272       3455       3261         0      1000   
2      0      0  ...      14331      14948      15549      1518      1500   
3      0      0  ...      28314      28959      29547      2000      2019   
4     -1      0  ...      20940      19146      19131      2000     36681   

   PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0         0         0         0         0                           1  
1      1000      1000         0      2000                           1  
2      1000      1000      1000      5000                           0  
3      1200      1100      1069      1000                           0  
4     10000      9000       689       679                           0  

[5 rows x 25 columns]

Verificando agora o número de linhas e colunas do dataset:

In [4]:
df.shape

(30000, 25)

Já sabemos que  dataset contém 30.000 linhas e 25 colunas, agora iniciaremos uma melhor exploração dos dados, verificando o nome das colunas.

In [5]:
df.columns

Index(['ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_1',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'default payment next month'],
      dtype='object')

De acordo com o nome das colunas, podemos identificar a coluna ID, que serve (ou deveria servir) como chave única de identificação para cada cliente, com o código a seguir podemos olhar detalhadamente a coluna ID e fazer uma contagem das chaves únicas.

In [6]:
df['ID'].nunique()

29687

Ou seja, identificamos que em um dataset de 30.000 linhas, temos um número menor de ID's únicos, o que nos mostra que alguns deles estão duplicando.

Podemos fazer uma contagem desses ID's e agrupando em uma variável definida como id_counts de forma a identificar quais estão aparecendo mais de uma vez.

In [7]:
id_counts = df['ID'].value_counts()
id_counts.head()

ad23fe5c-7b09    2
b46f955f-9e8a    2
5deff6b6-62ff    2
bf23c70b-bf12    2
c3ddce11-35e2    2
Name: ID, dtype: int64

Com o comando executado acima .head() nós conseguimos visualizar por padrão os 5 primeiros casos de ID's duplicados, agora iremos verificr e agrupar o número de entradas duplicadas de nosso dataset.

In [8]:
id_counts.value_counts()

1    29374
2      313
Name: ID, dtype: int64

Com essa checagem conseguimos verificar que a maioria dos ID's aparecem apenas uma única vez, enquanto que 313 deles aparecem duas vezes, nenhum apareceu mais de duas vezes.

Faremos a importação do numpy agora para continuarmos com a verificação e validação dos dados.

In [9]:
import numpy as np

Com numpy já carregado nós queremos agora apenas os ID's duplicados, que iremos agrupar em uma variável denominada dupe_mask

In [10]:
dupe_mask = id_counts == 2
dupe_mask[0:5]

ad23fe5c-7b09    True
b46f955f-9e8a    True
5deff6b6-62ff    True
bf23c70b-bf12    True
c3ddce11-35e2    True
Name: ID, dtype: bool

Essa variável é a máscara lógica criada para armazenar valores booleanos (True/False) - no caso como escolhemos id_counts == 2 
então ela deve nos retornar apenas valores True.

Verificando os dados de outra forma:

In [11]:
id_counts.index[0:5]

Index(['ad23fe5c-7b09', 'b46f955f-9e8a', '5deff6b6-62ff', 'bf23c70b-bf12',
       'c3ddce11-35e2'],
      dtype='object')

Iremos selecionar e armazenar os ID's duplicados em uma nova variável chamada dupe_ids
Depois iremos converter em uma lista e exibir seu tamanho.

In [12]:
dupe_ids = id_counts.index[dupe_mask]
dupe_ids = list(dupe_ids)
len(dupe_ids)

313

Verificando as 5 primeiras entradas - você pode alterar o número dentro do parênteses para exibir quantas entradas lhe for mais conveniente.

In [13]:
dupe_ids[0:5]

['ad23fe5c-7b09',
 'b46f955f-9e8a',
 '5deff6b6-62ff',
 'bf23c70b-bf12',
 'c3ddce11-35e2']

Usaremos o método .isin e .loc nas três primeiras entradas de nossas duplicadas, com isso iremos criar uma máscara booleana dos argumentos de todas as colunas, estaremos basicamente filtrando as três primeiras duplicadas do dataset e aplicando um método que nos permite visualizar se há algo de errado visualmente nos dados carregados.

In [14]:
df.loc[df['ID'].isin(dupe_ids[0:3]),:].head(10)

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
13146  5deff6b6-62ff     140000    2          2         2   32    -1     -1   
13246  5deff6b6-62ff          0    0          0         0    0     0      0   
26797  b46f955f-9e8a     280000    1          1         1   34    -1     -1   
26897  b46f955f-9e8a          0    0          0         0    0     0      0   
29685  ad23fe5c-7b09      50000    1          3         1   32     0      0   
29785  ad23fe5c-7b09          0    0          0         0    0     0      0   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
13146     -1     -2  ...          0       1186          0      3134         0   
13246      0      0  ...          0          0          0         0         0   
26797     -1     -2  ...        -21      15237       7345        76         0   
26897      0      0  ...          0          0          0         0         0   
29685      0      0  ...      12882       8131       3983      3000      2871   
29785      0      0  ...          0          0          0         0         0   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
13146         0      1186         0      1047                           1  
13246         0         0         0         0                           0  
26797         0     15258        36         0                           0  
26897         0         0         0         0                           0  
29685      1000       163      3983      3771                           1  
29785         0         0         0         0                           0  

[6 rows x 25 columns]

Facilmente conseguimos observar que durante a extração dos dados via SQL, ocorreu uma duplicação de alguns ID's e um deles tem uma linha somente com dados zerados.

Acredito que todos chegamos a mesma conclusão de que precisamos eliminar os dados que contém linhas zeradas, excetuando-se a coluna com ID, dessa forma eliminaremos o problema de duplicação.

Se criarmos uma matriz booleana do mesmo tamanho do DataFrame utilizando a condição == 0 poderemos agrupar esses ID's zerados.

In [15]:
df_zero_mask =df == 0

Vamos utilizar novamente o .loc para criarmos a máscara booleana e retornar o True, ou seja, os valores zerados a partir da segunda coluna, que são os dados que nos interessam.

In [16]:
feature_zero_mask = df_zero_mask.iloc[:,1:].all(axis=1)

In [17]:
sum(feature_zero_mask)

315

Vamos limpar o DataFrame agora excluindo as linhas que contém apenas zeros.

In [18]:
df_clean_1 = df.loc[~feature_zero_mask,:].copy()

Na verdade, criamos um novo DataFrame com o código anterior, df_clean_1, nele, nós excluimos tudo que era diferente de zero usando o operador lógico not, ~,e a resposta : para selecionar todas as colunas.

Verificando o shape desse novo DataFrame.

In [19]:
df_clean_1.shape

(29685, 25)

Agora constatando quantos desses ID's são únicos.

In [20]:
df_clean_1['ID'].nunique()

29685

Primeira etapa de verificação dos dados antes de iniciarmos nossa análise preditiva está concluída, conseguimos limpar com sucesso nosso DataFrame e temos agora apenas valores exclusivos nos ID's.

# EXPLORANDO E LIMPANDO OS DADOS

Após a primeira limpeza de dados, vamos agora checar os tipos de dados que compõem nosso dataset utilizando o info()

In [21]:
df_clean_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29685 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   ID                          29685 non-null  object
 1   LIMIT_BAL                   29685 non-null  int64 
 2   SEX                         29685 non-null  int64 
 3   EDUCATION                   29685 non-null  int64 
 4   MARRIAGE                    29685 non-null  int64 
 5   AGE                         29685 non-null  int64 
 6   PAY_1                       29685 non-null  object
 7   PAY_2                       29685 non-null  int64 
 8   PAY_3                       29685 non-null  int64 
 9   PAY_4                       29685 non-null  int64 
 10  PAY_5                       29685 non-null  int64 
 11  PAY_6                       29685 non-null  int64 
 12  BILL_AMT1                   29685 non-null  int64 
 13  BILL_AMT2                   29685 non-null  in

Vemos que das 25 colunas, temos duas que não estão com os dados classificados como integer, são elas ID, que já vimos, e Pay_1, que vamos olhar agora mais detalhamenta.

In [22]:
df_clean_1['PAY_1'].head(5)

0     2
1    -1
2     0
3     0
4    -1
Name: PAY_1, dtype: object

Os dados da coluna PAY_1 são mostrados a direita, vemos que temos dados documentados como 0, que não estão documentatos no dicionário dos dados.

Vamos prosseguir com uma contagem de valores da coluna PAY_1 para verificar os dados com maiores detalhes.

In [23]:
df_clean_1['PAY_1'].value_counts()

0                13087
-1                5047
1                 3261
Not available     3021
-2                2476
2                 2378
3                  292
4                   63
5                   23
8                   17
6                   11
7                    9
Name: PAY_1, dtype: int64

Conseguimos agora verificar o porquê desses dados estarem constando como objetc ao invés de int64, temos além de dados 0 e -2, uma string marcada como "Not available", ou seja, precisaremos lidar com esse fato mais tarde, há diversas formas de lidarmos com essa questão, nesse momento vamos eliminá-los de nosso dataset, esse assunto será retomado posteriormente com uma diferente abordagem.

Vamos ciar uma máscara lógica com os dados de PAY_1 descartando os dados ausentes, utilizaremos o operador != (que significa "diferente de" em Python)

In [24]:
valid_pay_1_mask = df_clean_1['PAY_1'] != 'Not available'

valid_pay_1_mask[0:5]

0    True
1    True
2    True
3    True
4    True
Name: PAY_1, dtype: bool

In [25]:
sum(valid_pay_1_mask)

26664